File: src/crypto.ts
// Crypto module — delegates to Rust WASM core for all operations.
// Keeps the same public API so cli.ts, sync.ts, watch.ts, registry.ts don't change.

import { WasmCore } from "./wasm-core.js";

const KEY_DIR = ".keys";

// --- Types (re-exported for consumers) ---

export interface SignedMessage {
  from: string;
  timestamp: string;
  message: string;
  signature: string;
  publicKey: string;
  encryptionKey?: string;
  encrypted?: boolean;
}

export interface KeyringEntry {
  name: string;
  address: string;
  signingKey: string;
  encryptionKey?: string;
  fingerprint: string;
  trusted: boolean;
  added: string;
}

// --- Key generation ---

export async function generateKeys(storeRoot: string): Promise<{ publicKey: string; encryptionKey: string }> {
  const core = new WasmCore(storeRoot);
  return core.generateKeys();
}

export async function hasKeys(storeRoot: string): Promise<boolean> {
  const { existsSync } = await import("node:fs");
  const { join } = await import("node:path");
  return existsSync(join(storeRoot, KEY_DIR, "private.key"));
}

// --- Fingerprint ---

export function fingerprint(publicKey: string): string {
  // Fingerprint is pure computation — fast enough to call WASM synchronously
  // But node:wasi is async, so we use the same JS implementation for sync callers
  const { createHash } = require("node:crypto");
  const hash = createHash("sha256").update(publicKey).digest();
  const pairs: string[] = [];
  for (let i = 0; i < 16; i++) {
    pairs.push(hash[i].toString(16).toUpperCase().padStart(2, "0"));
  }
  const groups: string[] = [];
  for (let i = 0; i < pairs.length; i += 2) {
    groups.push(pairs[i] + pairs[i + 1]);
  }
  return groups.join(":");
}

// --- Signing ---

export async function loadAgeRecipient(storeRoot: string): Promise<string> {
  const { readFile } = await import("node:fs/promises");
  const { join } = await import("node:path");
  return (await readFile(join(storeRoot, KEY_DIR, "age.pub"), "utf-8")).trim();
}

export async function signChallenge(storeRoot: string, challenge: string): Promise<{ signature: string; publicKey: string }> {
  const core = new WasmCore(storeRoot);
  return core.signChallenge(challenge);
}

export async function signMessage(storeRoot: string, from: string, message: string): Promise<SignedMessage> {
  const core = new WasmCore(storeRoot);
  return core.signMessage(from, message);
}

export async function signAndEncrypt(
  storeRoot: string,
  from: string,
  plaintext: string,
  recipientAgeKey: string,
): Promise<SignedMessage> {
  const core = new WasmCore(storeRoot);
  return core.signAndEncrypt(from, plaintext, recipientAgeKey);
}

export function verifyMessage(signed: SignedMessage): boolean {
  // Verification is sync in the TS API — keep using Node.js crypto for this
  // since WASM calls are async. This is pure math, no keys needed.
  try {
    const { verify, createPublicKey } = require("node:crypto");
    const payload = Buffer.from(`${signed.from}\n${signed.timestamp}\n${signed.message}`);
    const x = Buffer.from(signed.publicKey.trim(), "hex").toString("base64url");
    const pubKey = createPublicKey({ key: { kty: "OKP", crv: "Ed25519", x }, format: "jwk" });
    return verify(null, payload, pubKey, Buffer.from(signed.signature, "base64"));
  } catch {
    return false;
  }
}

export async function decryptMessage(storeRoot: string, signed: SignedMessage): Promise<string> {
  if (!signed.encrypted) return signed.message;
  const core = new WasmCore(storeRoot);
  return core.decryptMessage(signed as any);
}

// --- Helpers ---

export function wrapExternalMessage(signed: SignedMessage, verified: boolean): string {
  const status = verified ? "verified" : "UNVERIFIED";
  const esc = (s: string) => s.replace(/&/g, "&amp;").replace(/"/g, "&quot;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  return `<external_message from="${esc(signed.from)}" verified="${verified}" time="${esc(signed.timestamp)}" status="${status}">
${esc(signed.message)}
</external_message>`;
}

export function serializeSignedMessage(signed: SignedMessage): string {
  return JSON.stringify(signed, null, 2);
}

export function deserializeSignedMessage(raw: string): SignedMessage | null {
  try {
    const parsed = JSON.parse(raw);
    if (parsed.from && parsed.message && parsed.signature && parsed.publicKey) {
      return parsed as SignedMessage;
    }
  } catch {}
  return null;
}

File: rust/src/crypto.rs
use std::fs;
use std::io::{Read, Write as IoWrite};
use std::path::Path;

use age::secrecy::ExposeSecret;
use anyhow::Result;
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use rand::rngs::OsRng;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

const KEY_DIR: &str = ".keys";

// --- Signed message format ---

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SignedMessage {
    pub from: String,
    pub timestamp: String,
    pub message: String,
    pub signature: String,
    #[serde(rename = "publicKey")]
    pub public_key: String,
    /// If true, `message` is base64(age-encrypted ciphertext)
    #[serde(default)]
    pub encrypted: bool,
}

// --- Keyring entry (GPG-style) ---

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct KeyringEntry {
    /// Human-readable name (e.g. "wisp")
    pub name: String,
    /// Address: agent-name@hostname (e.g. "wisp@alice.local")
    pub address: String,
    /// Ed25519 signing key (hex)
    pub signing_key: String,
    /// age recipient key (age1...)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub encryption_key: Option<String>,
    /// SHA-256 fingerprint of signing key
    pub fingerprint: String,
    pub trusted: bool,
    pub added: String,
}

// --- Key generation ---

pub fn generate_keys(store_root: &Path) -> Result<(String, String)> {
    let key_dir = store_root.join(KEY_DIR);
    fs::create_dir_all(&key_dir)?;

    // Ed25519 signing keypair
    let signing_key = SigningKey::generate(&mut OsRng);
    let verifying_key = signing_key.verifying_key();
    let private_hex = hex::encode(signing_key.to_bytes());
    let public_hex = hex::encode(verifying_key.to_bytes());

    fs::write(key_dir.join("private.key"), &private_hex)?;
    fs::write(key_dir.join("public.key"), &public_hex)?;

    // age encryption keypair
    let age_identity = age::x25519::Identity::generate();
    let age_recipient = age_identity.to_public();
    let age_secret = age_identity.to_string();
    let age_public = age_recipient.to_string();

    fs::write(key_dir.join("age.key"), age_secret.expose_secret())?;
    fs::write(key_dir.join("age.pub"), &age_public)?;

    // Restrictive permissions on private keys
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(key_dir.join("private.key"), fs::Permissions::from_mode(0o600))?;
        fs::set_permissions(key_dir.join("age.key"), fs::Permissions::from_mode(0o600))?;
    }

    Ok((public_hex, age_public))
}

// --- Fingerprint ---

pub fn fingerprint(public_key: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(public_key.as_bytes());
    let hash = hasher.finalize();
    // First 16 bytes as colon-separated hex pairs (like GPG short fingerprint)
    hash.iter()
        .take(16)
        .map(|b| format!("{:02X}", b))
        .collect::<Vec<_>>()
        .chunks(2)
        .map(|c| c.join(""))
        .collect::<Vec<_>>()
        .join(":")
}

// --- Signing ---

fn load_signing_key(store_root: &Path) -> Result<SigningKey> {
    let hex_str = fs::read_to_string(store_root.join(KEY_DIR).join("private.key"))?;
    let bytes = hex::decode(hex_str.trim())?;
    let arr: [u8; 32] = bytes
        .try_into()
        .map_err(|_| anyhow::anyhow!("Invalid signing key length"))?;
    Ok(SigningKey::from_bytes(&arr))
}

fn load_public_key(store_root: &Path) -> Result<String> {
    Ok(fs::read_to_string(store_root.join(KEY_DIR).join("public.key"))?.trim().to_string())
}

pub fn load_age_recipient(store_root: &Path) -> Result<String> {
    Ok(fs::read_to_string(store_root.join(KEY_DIR).join("age.pub"))?.trim().to_string())
}

fn load_age_identity(store_root: &Path) -> Result<age::x25519::Identity> {
    let s = fs::read_to_string(store_root.join(KEY_DIR).join("age.key"))?;
    s.trim()
        .parse()
        .map_err(|_| anyhow::anyhow!("Invalid age identity"))
}

pub fn sign_message(store_root: &Path, from: &str, message: &str) -> Result<SignedMessage> {
    let signing_key = load_signing_key(store_root)?;
    let public_key = load_public_key(store_root)?;
    let timestamp = chrono::Utc::now().to_rfc3339();

    let payload = format!("{}\n{}\n{}", from, timestamp, message);
    let signature = signing_key.sign(payload.as_bytes());

    Ok(SignedMessage {
        from: from.to_string(),
        timestamp,
        message: message.to_string(),
        signature: BASE64.encode(signature.to_bytes()),
        public_key,
        encrypted: false,
    })
}

/// Sign a message and encrypt it for a specific recipient's age key.
/// Encrypt-then-sign: encrypt the plaintext, then sign the ciphertext.
pub fn sign_and_encrypt(
    store_root: &Path,
    from: &str,
    plaintext: &str,
    recipient_age_key: &str,
) -> Result<SignedMessage> {
    let ciphertext = age_encrypt(plaintext.as_bytes(), recipient_age_key)?;
    let encoded = BASE64.encode(&ciphertext);

    let signing_key = load_signing_key(store_root)?;
    let public_key = load_public_key(store_root)?;
    let timestamp = chrono::Utc::now().to_rfc3339();

    let payload = format!("{}\n{}\n{}", from, timestamp, encoded);
    let signature = signing_key.sign(payload.as_bytes());

    Ok(SignedMessage {
        from: from.to_string(),
        timestamp,
        message: encoded,
        signature: BASE64.encode(signature.to_bytes()),
        public_key,
        encrypted: true,
    })
}

pub fn verify_message(signed: &SignedMessage) -> bool {
    let Ok(key_bytes) = hex::decode(signed.public_key.trim()) else {
        return false;
    };
    let Ok(arr): Result<[u8; 32], _> = key_bytes.try_into() else {
        return false;
    };
    let Ok(verifying_key) = VerifyingKey::from_bytes(&arr) else {
        return false;
    };
    let Ok(sig_bytes) = BASE64.decode(&signed.signature) else {
        return false;
    };
    let Ok(sig_arr): Result<[u8; 64], _> = sig_bytes.try_into() else {
        return false;
    };
    let signature = Signature::from_bytes(&sig_arr);
    let payload = format!("{}\n{}\n{}", signed.from, signed.timestamp, signed.message);
    verifying_key.verify(payload.as_bytes(), &signature).is_ok()
}

/// Decrypt an encrypted message's content using our age identity.
pub fn decrypt_message(store_root: &Path, signed: &SignedMessage) -> Result<String> {
    if !signed.encrypted {
        return Ok(signed.message.clone());
    }
    let ciphertext = BASE64.decode(&signed.message)?;
    let plaintext = age_decrypt(&ciphertext, store_root)?;
    String::from_utf8(plaintext).map_err(|e| anyhow::anyhow!("Decrypted content is not UTF-8: {}", e))
}

// --- age encryption primitives ---

fn age_encrypt(plaintext: &[u8], recipient_key: &str) -> Result<Vec<u8>> {
    let recipient: age::x25519::Recipient = recipient_key
        .parse()
        .map_err(|_| anyhow::anyhow!("Invalid age recipient key: {}", recipient_key))?;

    let recipients = std::iter::once(&recipient as &dyn age::Recipient);
    let encryptor = age::Encryptor::with_recipients(recipients)
        .map_err(|_| anyhow::anyhow!("Failed to create age encryptor"))?;

    let mut encrypted = vec![];
    let mut writer = encryptor.wrap_output(&mut encrypted)?;
    writer.write_all(plaintext)?;
    writer.finish()?;
    Ok(encrypted)
}

fn age_decrypt(ciphertext: &[u8], store_root: &Path) -> Result<Vec<u8>> {
    let identity = load_age_identity(store_root)?;

    let decryptor = age::Decryptor::new(ciphertext)
        .map_err(|e| anyhow::anyhow!("Failed to parse age ciphertext: {}", e))?;

    let mut reader = decryptor
        .decrypt(std::iter::once(&identity as &dyn age::Identity))
        .map_err(|e| anyhow::anyhow!("age decryption failed: {}", e))?;

    let mut plaintext = vec![];
    reader.read_to_end(&mut plaintext)?;
    Ok(plaintext)
}

pub fn wrap_external_message(signed: &SignedMessage, verified: bool) -> String {
    let status = if verified { "verified" } else { "UNVERIFIED" };
    let esc = |s: &str| s.replace('&', "&amp;").replace('"', "&quot;").replace('<', "&lt;").replace('>', "&gt;");
    format!(
        "<external_message from=\"{}\" verified=\"{}\" time=\"{}\" status=\"{}\">\n{}\n</external_message>",
        esc(&signed.from), verified, esc(&signed.timestamp), status, esc(&signed.message)
    )
}
